test/extended/util/framework.go

```diff
@@ -56,18 +60,99 @@ func WaitForInternalRegistryHostname(oc *CLI) (string, error) {
+
+		// now confirm the OCM was restarted after the image registry was set;
+		// to do that, we are going to compare the create time of the ocm pods
+		// with the last transition times of the ocm operator object progressing condition;
+		// we'll also confirm timings between the OCM rollout and image config creation,
+		// and we'll verify that the OCM config's internal registry hostname matches
+		// the image config's internal registry hostname
+		// get the OCM cfg
+		ocm, err := oc.AdminOperatorClient().OperatorV1().OpenShiftControllerManagers().Get("cluster", metav1.GetOptions{})
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return false, nil
+			}
+			return false, err
+		}
+		// get the OCM observed config object to verify it has the right internal registry hostname
+		observedConfig := map[string]interface{}{}
+		err = json.Unmarshal(ocm.Spec.ObservedConfig.Raw, &observedConfig)
+		if err != nil {
+			return false, nil
+		}
+		internalRegistryHostnamePath := []string{"dockerPullSecret", "internalRegistryHostname"}
+		currentRegistryHostname, _, err := unstructured.NestedString(observedConfig, internalRegistryHostnamePath...)
+		if err != nil {
+			e2e.Logf("error procesing observed config %#v", err)
+			return false, nil
+		}
+		if currentRegistryHostname != registryHostname {
+			e2e.Logf("OCM observed config hostname %s does not match image config hostname %s", currentRegistryHostname, registryHostname)
+			return false, nil
+		}
+		// find oldest pod
+		pods, err := oc.AdminKubeClient().CoreV1().Pods("openshift-controller-manager").List(metav1.ListOptions{})
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return false, nil
+			}
+			return false, err
+		}
+		oldestPod := metav1.Time{}
+		for _, pod := range pods.Items {
+			if oldestPod.Time.IsZero() {
+				oldestPod = pod.CreationTimestamp
+				e2e.Logf("oldest OCM pod is %s with time %s", pod.Name, pod.CreationTimestamp.String())
+				continue
+			}
+			if !oldestPod.Before(&pod.CreationTimestamp) {
+				e2e.Logf("oldest OCM pod is %s with time %s", pod.Name, pod.CreationTimestamp.String())
+				oldestPod = pod.CreationTimestamp
+			}
+		}
+		// now cycle through the OCM operator conditions and make sure the Progressing condition is done and is older than the OCM pods
+		for _, condition := range ocm.Status.Conditions {
+			if condition.Type != operatorv1.OperatorStatusTypeProgressing {
+				continue
+			}
+			if condition.Status != operatorv1.ConditionFalse {
+				e2e.Logf("OCM rollout still progressing or in error: %v", condition.Status)
+				return false, nil
+			}
+			// make sure OCM rolled out after image config created; since the image registry operator sets the internal
+			// registry hostname when it creates the image config object, then making sure the latest OCM rollout is
+			// after its create time is another indication that config value was processed by the OCM
+			if condition.LastTransitionTime.Before(&imageConfig.CreationTimestamp) {
```

Comment:
i don't know that looking at the imageconfig creationtimestamp is meaningful.  Pretty sure we create that object when the cluster is created (even before we populate it w/ the internal registry hostname).

I think that's going to be the most difficult part of this approach....  being confident that a given controller pod is "newer" than the time at which the internal registry hostname config value became available. Because there's no specific timestamp for that value being set in a config resource.

Reply:
My initial thought as well re: image config create time stamp.

But in my prototyping, the image config was always at a generation of 1 though, and in the image registry operator, I see that on the create of the image.config object, they are setting the internal registry hostname.

```
apiVersion: config.openshift.io/v1
kind: Image
metadata:
  annotations:
    release.openshift.io/create-only: "true"
  creationTimestamp: "2019-08-12T20:19:52Z"
  generation: 1
  name: cluster
  resourceVersion: "8460"
  selfLink: /apis/config.openshift.io/v1/images/cluster
  uid: 8abeb8c5-bd3e-11e9-a339-02e2cbcb7c6e
spec: {}
status:
  internalRegistryHostname: image-registry.openshift-image-registry.svc:5000
```

And here is the creation code, where they immediately set the status after the create:  https://github.com/openshift/cluster-image-registry-operator/blob/master/pkg/resource/imageconfig.go#L78-L108

I took those pieces to mean we could effectively leverage the create timestamp.

thoughts?

Reply:
that logic will take effect only if the resource does not already exist.  But it could already exist (in which case the registry operator just does an Update path), and in fact i'd expect it to already exist due to:

https://github.com/openshift/cluster-config-operator/blob/master/empty-resources/0000_05_config-operator_02_image.cr.yaml

it's possible it's non-deterministic who creates the image config resources (the cluster config operator, or the image registry operator).

it's also possible that generation value isn't being incremented properly, as I just added a route to my registry (which causes the image config resource to be updated w/ an external hostname) and the generation is still "1".  So i'd say that's not a reliable indicator that the creation state == current state of the resource.


In any case the image config object's timestamp isn't really what matters, it's the controller manager's config timestamp.  You need to know that all the controller pods are newer than the time at which the controller manager operator config resource was updated to include the internal registry hostname (which it gets from the image config resource).  Which again I think will be hard to determine since you'd have to know not the creation timestamp but specifically when a particular field was updated.

Reply:
OK then ... punting on correlating creation times of the OCM pods against any of the config 

I propose leaving in:
- at least making sure the OCM config's setting matches the image config's setting (in addition to the current check that the image config is set)
- leave in the check to make sure an OCM rollout is not in progress (including all 3 pods properly recreted) .... current e2e results already show that an OCM rollout is still in progress when the e2e-aws-build run starts .... I've seen this in other recent test exercises ... OCM progressing goes to false before the daemonset/pod rollout finishes ... I was reluctant to change that earlier, but maybe something to revisit.... I'm referring to my recent changes in https://github.com/openshift/origin/blob/master/test/extended/builds/cluster_config.go as part of moving integration tests to extended tests
- either contrive log parsing based on the current log noted above, or add new logging that makes said parsing easier ... and use that to confirm the config has been consumed

- or would adding an annotation for last update time on the OCM config be too rebellious?

thoughts?

Reply:
> at least making sure the OCM config's setting matches the image config's setting (in addition to the current check that the image config is set)

makes sense

> leave in the check to make sure an OCM rollout is not in progress (including all 3 pods properly recreted) 

makes sense, and may well be "sufficient" to resolve the core issue here.

> OCM progressing goes to false before the daemonset/pod rollout finishes

really?  is that a bug?  it sounds wrong.

> either contrive log parsing based on the current log noted above, or add new logging that makes said parsing easier ... and use that to confirm the config has been consumed

per my earlier suggestion i certainly support this one :)


> or would adding an annotation for last update time on the OCM config be too rebellious?

i'm pretty sure that's been done elsewhere, so not too rebellious per se.

Reply:
> > at least making sure the OCM config's setting matches the image config's setting (in addition to the current check that the image config is set)
> 
> makes sense

cool

> 
> > leave in the check to make sure an OCM rollout is not in progress (including all 3 pods properly recreted)
> 
> makes sense, and may well be "sufficient" to resolve the core issue here.

that thought occurred to me as well, though seemed perhaps hopeful at first blush

> 
> > OCM progressing goes to false before the daemonset/pod rollout finishes
> 
> really? is that a bug? it sounds wrong.

@adambkaplan and I were of the same thinking back with the integration test to extended test work, but I have not had time to research the history there and reach out to whomever touched that piece last.

I'm more of a mind at the moment to handle that separately.

I've opened https://bugzilla.redhat.com/show_bug.cgi?id=1740887 to track.
> 
> > either contrive log parsing based on the current log noted above, or add new logging that makes said parsing easier ... and use that to confirm the config has been consumed
> 
> per my earlier suggestion i certainly support this one :)
> 
> > or would adding an annotation for last update time on the OCM config be too rebellious?
> 
> i'm pretty sure that's been done elsewhere, so not too rebellious per se.

OK cool I'll add that as well, though will take on log parsing initially.  Based if/how it lands, if possible I'll nuke the log parsing once it is ready.